### PR TITLE
Migrate to Ktor 2

### LINKED ISSUE #47: Crash: Using blocking primitives on this dispatcher is not allowed
MoshiConverter crashes with: "Using blocking primitives on this dispatcher is not allowed. Consider using async channel instead or use blocking primitives in withContext(Dispatchers.IO) instead."

What you need to do is:
```
    return withContext(Dispatchers.IO) {
      val source = channel.toInputStream().source().buffer()
      val type = request.type
      moshi.adapter(type.javaObjectType).fromJson(source)
    }
```
In the MoshiConverter

### LINKED ISSUE #48: Support ktor 2.0
in ktor 2.0 some of the package names have changed ([available thru the EAP](https://ktor.io/eap/))

ktor 1.x: 
```
io.ktor.features.ContentNegotiation
```

ktor 2.x: 
```
io.ktor.server.plugins.ContentNegotiation
```

### PR DESCRIPTION
**Pull request by ianprime0509**
This commit migrates the project to Ktor 2.

In Ktor 2, `JsonFeature` is deprecated and superseded by a client `ContentNegotiation` plugin, which results in the core serialization logic being shared between client and server:
https://ktor.io/docs/migrating-2.html#serialization-client Hence, as part of this migration, the client and server modules are combined into a single module which has no client-specific or server-specific dependencies.

Several other upgrades (Kotlin, Gradle, and Moshi) are included in this change, as they were directly or indirectly required for the Ktor 2 upgrade.

Closes #47
Closes #48

### CODE REVIEW

File: core/src/main/kotlin/com/hypercubetools/ktor/moshi/MoshiConverter.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import io.ktor.utils.io.jvm.javaio.toInputStream
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okio.buffer
+import okio.source
+
+class MoshiConverter(private val moshi: Moshi = Moshi.Builder().build()) : ContentConverter {
+    override suspend fun deserialize(charset: Charset, typeInfo: TypeInfo, content: ByteReadChannel): Any? {
+        try {
+            return withContext(Dispatchers.IO) {
```
Comment (ianprime0509) [inline core/src/main/kotlin/com/hypercubetools/ktor/moshi/MoshiConverter.kt:25]: This is what fixes #47. The same pattern is also used in the official `ContentConverter` implementations, e.g. https://github.com/ktorio/ktor/blob/d300441cb70b6a590ebe263febc36f3fe825b4f6/ktor-shared/ktor-serialization/ktor-serialization-gson/jvm/src/GsonConverter.kt#L54

File: core/src/main/kotlin/com/hypercubetools/ktor/moshi/MoshiConverter.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                val source = content.toInputStream().source().buffer()
+                moshi.adapter<Any?>(typeInfo.reifiedType).fromJson(source)
+            }
+        } catch (e: JsonDataException) {
+            throw JsonConvertException("Unable to convert JSON", e)
```
Comment (ianprime0509) [inline core/src/main/kotlin/com/hypercubetools/ktor/moshi/MoshiConverter.kt:30]: The official `ContentConverter` implementations all wrap library-specific deserialization exceptions in `JsonConvertException`, so this change does the same thing here as well.

File: core/src/test/kotlin/com/hypercubetools/ktor/moshi/MoshiConverterTest.kt [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+const val SOME_NAME = "Foosius"
+val SOME_FOO = Foo(id = SOME_ID, name = SOME_NAME)
+const val SOME_FOO_JSON = """{"id":$SOME_ID,"name":"$SOME_NAME"}"""
+const val SOME_KEY = "bar-123"
+const val SOME_COUNT = 50
+val SOME_BAR = Bar(key = SOME_KEY, count = SOME_COUNT)
+const val SOME_BAR_JSON = """{"key":"$SOME_KEY","count":$SOME_COUNT}"""
+val SOME_NOT_JSON_TYPE = ContentType.Text.Plain
+
+class MoshiConverterTest {
```
Comment (ianprime0509) [inline core/src/test/kotlin/com/hypercubetools/ktor/moshi/MoshiConverterTest.kt:59]: This is pretty much just a combination of the existing server and client tests, included in two inner classes. I added a couple more client tests to handle the new exception handling cases that were added.

**Comment by ianprime0509** [comment]
Sorry about all the pushes; I discovered a few things after I created the PR which I had to fix in the CI configuration and certain Gradle plugins due to the module consolidation and Gradle update.

**Comment by plannigan** [comment]
Sorry for not responding earlier. Somehow I was only getting notified about mentions. I have some time now that will allow me to review this and get this project back into shape.

**Comment by plannigan** [comment]
Thanks so much for doing this work. It seems like the only "issue" with the PR is the failing coverage checks. That config can [specify a threshold](https://docs.codecov.com/docs/commit-status#threshold) to allow for the coverage to lower slightly ([example](https://github.com/plannigan/hyper-bump-it/blob/b6c8aaf2bdd9ad8923e2ba739b855a506a573d81/.github/codecov.yml#L14-L16)). If you make that adjustment, the checks will pass and I'll merge the PR.

**Comment by ianprime0509** [comment]
No problem, happy to help! I followed a similar pattern for the updated `.codecov.yml` as in the example you shared; if you think the target or threshold values should be different from what I added, please let me know and I can update them.

**Review by plannigan — APPROVED** [review]

**Pull request merged**